ddtrace/profiling/collector/_lock.py

```diff
@@ -233,6 +235,47 @@ def __enter__(self, *args, **kwargs):
     def __exit__(self, *args, **kwargs):
         self._release(self.__wrapped__.__exit__, *args, **kwargs)
 
+    # Get lock acquire/release call location and variable name the lock is assigned to
+    def _get_lock_call_loc_with_name(self) -> typing.Optional[str]:
+        try:
+            # We expect the call stack to be like this:
+            # 0: this
+            # 1: _acquire/_release
+            # 2: acquire/release (or __enter__/__exit__)
+            # 3: caller frame
+            # And we expect addtioinal frame if WRAPT_C_EXT is False
+            frame = sys._getframe(3 if WRAPT_C_EXT else 4)
+            code = frame.f_code
+            call_loc = "%s:%d" % (os.path.basename(code.co_filename), frame.f_lineno)
+
+            var_name = None
+
+            # Search for global variables
+            for name, value in frame.f_globals.items():
+                if value == self:
+                    var_name = name
+                    break
+                for attribute in dir(value):
+                    if not attribute.startswith("__") and getattr(value, attribute) == self:
```

Comment:
Yeah, I think omitting `__`-prefixed attributes is pragmatic.  For a given object, in the typical case, _most_ of those will be inherited junk that won't be a lock (... until a class has a private lock we won't know about now), so this saves time.

One feature of this implementation is that we'll miss attributes of attributes.  In general, we'd have to do a recursive scan through the attribute tree in order to exhaust all possibilities.  I don't think that's worth doing, just working through the dynamics of the solution for my own benefit.

Reply:
Added a test case for a private lock, and it just works! 

re: recursive scan
Yeah, also added a test case to show that we can't find the name in this case. I also agree that the more worrying part is that we iterate through too many globals and some customers might be hit by that.